---
title: "Using SNPs"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using SNPs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(ncbi)
```

## Find gene names of membrane proteins

At the NCBI website (`https://www.ncbi.nlm.nih.gov/`),
click on the 'Gene' database, then search for 'Membrane protein', for the organism *Homo sapiens*, or use the API `https://www.ncbi.nlm.nih.gov/gene?term=(membrane%20protein)%20AND%20homo%20sapiens%5BOrganism%5D` to get 1123 hits.

![](find_gene_names.png)

Search for 

```{r}
# Do an NCBI search
membrane_protein_result <- rentrez::entrez_search(db = "Gene", term = "((membrane protein) AND Homo sapiens[ORGN]) AND alive[prop]")
Sys.sleep(1)

print(paste0("Found ", membrane_protein_result$count, " matches, from the query '", membrane_protein_result$QueryTranslation, "'"))
```
Here we find 1123 matches again. We did have to expand the query, by adding `AND alive[prop]` to only show the alive entries, a thing the web interface does by default.

Here we do the same, using the `get_all_human_membrane_protein_gene_ids` function. From there, we get the first six:

```{r}
# Get the membrane proteins' gene IDs
gene_ids <- head(get_all_human_membrane_protein_gene_ids())
gene_ids
```
From our gene IDs, we can get the gene names:

```{r}
gene_names <- get_gene_names_from_human_gene_ids(gene_ids)
```

## From a gene name, find a SNP

### Gene name without SNPS

In this example, we'll use the EFGR gene. There will be no (human) SNPs for that gene however, as we'll discover

### Gene name with SNPS

In this example, we'll use the TNF gene.

```{r}
gene_name <- gene_names[2]
testthat::expect_equal("TNF", gene_name)
```

On the NCBI website, use the SNP database and search for the gene name 'TNF', or use this API call:  `https://www.ncbi.nlm.nih.gov/snp/?term=TNF%5BGene%20Name%5D`

![](tnf_snps.png)

From R, we do:

```{r}
snp_ids <- get_snp_ids_from_gene_name(gene_name)
head(snp_ids)
```

## From a SNP, get the protein sequence and location

## From a SNP, that is not translated

We'll use the first SNP ID, 1583051968:

```{r}
untranslated_tnf_snp_result <- snp_ids[1]
testthat::expect_equal("1583051968", untranslated_tnf_snp_result)
```

To get the protein sequence using the NCBI website,
search for `1583051968` (or `rs1583051968`, the `rs` denotes
it's a SNP),
or use the API call \code{https://www.ncbi.nlm.nih.gov/snp/?term=1583051968}

![](snp_1583051968.png)

Clicking on the `rs` takes us to [https://www.ncbi.nlm.nih.gov/snp/rs1583051968](https://www.ncbi.nlm.nih.gov/snp/rs1583051968). Scrolling down gives the genomic context:

![](snp_1583051968_report.png)

As the report shows no green (nor red, nor blue) band, means that this SNP does not modify a translation product. Zooming out comfirms this:

![](snp_1583051968_report_not_on_protein.png)

## From a SNP, that is transcripted

We'll use the fifth SNP ID, 1583051188:

```{r}
transcripted_tnf_snp_result <- snp_ids[5]
testthat::expect_equal("1583051188", transcripted_tnf_snp_result)
```

At NCBI website, at [https://www.ncbi.nlm.nih.gov/snp/rs1583051188](https://www.ncbi.nlm.nih.gov/snp/rs1583051188) we can see it is transcribed to mRNA (but not to protein):
![](snp_1583051188_report.png)


## From a SNP, that is transcripted and translated into indentical protein

We'll use the sevent SNP ID, 1583050033:

```{r}
translated_tnf_snp_result <- snp_ids[7]
testthat::expect_equal("1583050033", translated_tnf_snp_result)
```

At [https://www.ncbi.nlm.nih.gov/snp/rs1583050033](https://www.ncbi.nlm.nih.gov/snp/rs1583050033) one can see there is a protein

![](snp_1583050033_report.png)

Hovering over the protein (that is, the red bar), we see
that the protein is called `NP_000585.2` and that our
SNP acts on the 196 amino acid.

![](snp_1583050033_report_zoom.png)

Now, to do the same thing from R:

```{r}
hgvs <- get_snp_variations_in_protein(translated_tnf_snp_result)
tryCatch(is_hgvs_in_tmh(hgvs), error = function(e) print(e))
```
The error message is clear: the SNP did not cause a mutation in the protein.

## From a SNP, that is transcripted and translated into indentical protein

We'll use the eight SNP ID, X1583050033:

```{r}
translated_tnf_snp_result <- snp_ids[8]
testthat::expect_equal("1583049783", translated_tnf_snp_result)
```

At [https://www.ncbi.nlm.nih.gov/snp/rs1583049783](https://www.ncbi.nlm.nih.gov/snp/rs1583049783) one can see there is a protein

![](snp_1583049783_report.png)

Hovering over the protein (that is, the red bar), we see
that the protein is called `NP_000585.2` and that our
SNP acts on the 144 amino acid.

![](snp_1583049783_report_zoom.png)

Now, to do the same thing from R:

```{r}
hgvs <- get_snp_variations_in_protein(translated_tnf_snp_result)
```

Here the protein is actually changed.

From this variation, we now measure where in the protein the mutation occurs:

```{r}
hgvs <- get_snp_variations_in_protein(translated_tnf_snp_result)
is_hgvs_in_tmh(hgvs)
```
